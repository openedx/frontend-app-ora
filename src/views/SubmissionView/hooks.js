import { useEffect, useReducer } from 'react';

import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

import { submitResponse, saveResponse, uploadFiles } from 'data/services/lms/hooks/actions';
import { MutationStatus } from 'data/services/lms/constants';

const useSubmissionViewHooks = () => {
  const submitResponseMutation = submitResponse();
  const saveResponseMutation = saveResponse();
  const uploadFilesMutation = uploadFiles();
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  const [submission, dispatchSubmission] = useReducer(
    (state, payload) => ({ ...state, isDirty: true, ...payload }),
    { ...pageData?.submission, isDirty: false },
  );

  useEffect(() => {
    // a workaround to update the submission state when the pageData changes
    if (pageData?.submission) {
      dispatchSubmission({ ...pageData.submission, isDirty: false });
    }
  }, [pageData?.submission]);

  const onTextResponseChange = (index) => (textResponse) => {
    dispatchSubmission({
      response: {
        ...submission.response,
        textResponses: [
          ...submission.response.textResponses.slice(0, index),
          textResponse,
          ...submission.response.textResponses.slice(index + 1),
        ],
      },
    });
  };

  const onFileUploaded = (args) => {
    const fileUploads = uploadFilesMutation.mutate(args);
    dispatchSubmission({ response: { ...submission.response, fileUploads } });
  };

  const submitResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    submitResponseMutation.mutate(submission);
  };

  const saveResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    saveResponseMutation.mutate(submission);
  };

  return {
    submitResponseHandler,
    submitResponseStatus: submitResponseMutation.status,
    saveResponseHandler,
    saveResponseStatus: saveResponseMutation.status,
    draftSaved: saveResponseMutation.status === MutationStatus.success && !submission.isDirty,
    pageData,
    oraConfigData,
    submission,
    dispatchSubmission,
    onTextResponseChange,
    onFileUploaded,
  };
};

export default useSubmissionViewHooks;
