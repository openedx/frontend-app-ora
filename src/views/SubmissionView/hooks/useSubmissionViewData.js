import React from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useGlobalState,
  useRubricConfig,
  useSubmitResponse,
  useSetHasSubmitted,
  useHasSubmitted,
  useRefreshPageData,
  useSetResponse,
  useResponse,
} from 'hooks/app';
import {
  useRefreshUpstream,
} from 'hooks/modal';
import { stepStates, stepNames } from 'constants/index';

import useTextResponsesData from './useTextResponsesData';
import useUploadedFilesData from './useUploadedFilesData';

const stateKeys = StrictDict({
  hasSavedDraft: 'hasSavedDraft',
});

const useSubmissionViewData = () => {
  const [hasSavedDraft, setHasSavedDraft] = useKeyedState(stateKeys.hasSavedDraft, false);
  const hasSubmitted = useHasSubmitted();
  const setHasSubmitted = useSetHasSubmitted();
  const submitResponseMutation = useSubmitResponse();
  const rubricConfig = useRubricConfig();
  const globalState = useGlobalState({ step: stepNames.submission });
  const refreshPageData = useRefreshPageData();
  const setResponse = useSetResponse();
  const response = useResponse();
  const refreshUpstream = useRefreshUpstream();
  const stepState = hasSubmitted ? stepStates.submitted : globalState.stepState;

  const {
    textResponses,
    onUpdateTextResponse,
    saveResponse,
    saveResponseStatus,
    finishLater,
    finishLaterStatus,
  } = useTextResponsesData({ setHasSavedDraft });

  const {
    uploadedFiles,
    onFileUploaded,
    onDeletedFile,
  } = useUploadedFilesData();

  const submitResponseHandler = React.useCallback(() => {
    submitResponseMutation.mutateAsync({
      textResponses,
      uploadedFiles,
    }).then(() => {
      setResponse({ textResponses, uploadedFiles });
      setHasSubmitted(true);
      refreshPageData();
      refreshUpstream();
    });
  }, [ // eslint-disable-line react-hooks/exhaustive-deps
    setHasSubmitted,
    submitResponseMutation,
    textResponses,
    uploadedFiles,
  ]);

  return {
    actionOptions: {
      finishLater,
      finishLaterStatus,
      saveResponse,
      saveResponseStatus,
      submit: submitResponseHandler,
      submitStatus: submitResponseMutation.status,
      hasSubmitted,
    },
    response: hasSubmitted
      ? response
      : { textResponses, uploadedFiles },
    hasSubmitted,
    onUpdateTextResponse,
    isDraftSaved: hasSavedDraft,
    onDeletedFile,
    onFileUploaded,
    showRubric: rubricConfig.showDuringResponse,
    isReadOnly: stepState === stepStates.done || hasSubmitted,
  };
};

export default useSubmissionViewData;
