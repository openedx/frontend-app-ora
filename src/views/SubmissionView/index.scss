@import "@edx/paragon/scss/core/core";

.assessment-content-layout {
  & > div.content-body {
    height: 100%;
    .row {
      height: 100%;
    }
  }
  width: fit-content;
  margin: auto;
  height: 100%;

  .content-wrapper {
    max-width: $max-width-lg;
  }
}

.ora-tinymce .tox-tinymce {
  background-color: $white;
}

@include media-breakpoint-down(sm) {
  .assessment-content-layout {
    .content-wrapper {
      width: 100%;
    }
  }
}

