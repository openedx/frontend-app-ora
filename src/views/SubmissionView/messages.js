import { defineMessages } from '@edx/frontend-platform/i18n';
import { MutationStatus } from 'constants/index';

const messages = defineMessages({
  yourResponse: {
    id: 'frontend-app-ora.SubmissionView.yourResponse',
    defaultMessage: 'Your response',
    description: 'Label for the response textarea',
  },
  draftSaved: {
    id: 'frontend-app-ora.SubmissionView.draftSaved',
    defaultMessage: 'Draft saved',
    description: 'Label for the draft saved message',
  },
  instructions: {
    id: 'frontend-app-ora.SubmissionView.instructions',
    defaultMessage: 'Instructions',
    description: 'Label for the instructions textarea',
  },
  instructionsText: {
    id: 'frontend-app-ora.SubmissionView.instructionsText',
    defaultMessage: `Create a response to the prompt below.
    Progress will be saved automatically and you can return to complete your
    progress at any time. After you submit your response, you cannot edit
    it.`,
    description: 'Description for the instructions textarea',
  },
  startTraining: {
    id: 'ora-grading.SubmissionView.startTraining',
    defaultMessage: 'Start practice grading',
    description: 'Action button text for start action to student training step',
  },
});

export const submitActionMessages = defineMessages({
  [MutationStatus.idle]: {
    id: 'ora-grading.SubmissionAction.submit',
    defaultMessage: 'Submit response',
    description: 'Submit button text',
  },
  [MutationStatus.loading]: {
    id: 'ora-grading.SubmissionAction.submitting',
    defaultMessage: 'Submitting response',
    description: 'Submit button text while submitting',
  },
  [MutationStatus.success]: {
    id: 'ora-grading.SubmissionAction.submitted',
    defaultMessage: 'Response submitted',
    description: 'Submit button text after successful submission',
  },

});

export const saveActionMessages = defineMessages({
  [MutationStatus.idle]: {
    id: 'ora-grading.SaveAction.save',
    defaultMessage: 'Save for later',
    description: 'Save for later button text',
  },
  [MutationStatus.loading]: {
    id: 'ora-grading.SaveAction.saving',
    defaultMessage: 'Saving response',
    description: 'Save for later button text while saving',
  },
});

export default {
  ...messages,
  ...submitActionMessages,
  ...saveActionMessages,
};
