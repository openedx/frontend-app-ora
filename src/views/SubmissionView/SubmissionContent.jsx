import React from 'react';

import { useORAConfigData } from 'data/services/lms/hooks/selectors';

import Prompt from 'components/Prompt';
import TextResponse from 'components/TextResponse';
import FileUpload from 'components/FileUpload';

const SubmissionContent = () => {
  const { prompts } = useORAConfigData();
  return (
    <div>
      {
        prompts.map((prompt, index) => (
          <div>
            <Prompt promptIndex={index} />
            <TextResponse promptIndex={index} />
          </div>
        ))
      }
      <FileUpload />
    </div>
  );
};

export default SubmissionContent;
