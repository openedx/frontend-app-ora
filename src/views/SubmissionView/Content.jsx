import React from 'react';
import PropTypes from 'prop-types';

import { Icon } from '@edx/paragon';
import { CheckCircle } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import {
  usePrompts,
  useSubmissionConfig,
  useStepState,
} from 'data/services/lms/hooks/selectors';
import { stepNames, stepStates } from 'data/services/lms/constants';

import FileUpload from 'components/FileUpload';
import Instructions from 'components/Instructions';
import Prompt from 'components/Prompt';
import TextResponse from 'components/TextResponse';
import TextResponseEditor from 'components/TextResponseEditor';
import StatusAlert from 'components/StatusAlert';

import messages from './messages';

const SubmissionContent = ({
  textResponses,
  uploadedFiles,
}) => {
  const submissionConfig = useSubmissionConfig();

  const stepState = useStepState({ step: stepNames.submission });
  const isReadOnly = stepState === stepStates.completed;
  const prompts = usePrompts();
  const { formatMessage } = useIntl();
  const createTextResponse = (index) => (isReadOnly
    ? <TextResponse response={textResponses.value[index]} />
    : (
      <TextResponseEditor
        submissionConfig={submissionConfig}
        value={textResponses.value[index]}
        onChange={textResponses.onChange(index)}
      />
    ));

  return (
    <div>
      <div className="d-flex justify-content-between">
        <h2 className="mb-4">{formatMessage(messages.yourResponse)}</h2>
        {(!isReadOnly && textResponses.draftSaved) && (
          <p className="d-flex text-primary">
            <Icon src={CheckCircle} />
            {formatMessage(messages.draftSaved)}
          </p>
        )}
      </div>

      <StatusAlert />
      <Instructions />

      {prompts.map((prompt, index) => (
        // eslint-disable-next-line react/no-array-index-key
        <div key={index}>
          <Prompt prompt={prompt} />
          {createTextResponse(index)}
        </div>
      ))}

      <FileUpload
        onDeletedFile={uploadedFiles.onDeletedFile}
        onFileUploaded={uploadedFiles.onFileUploaded}
        uploadedFiles={uploadedFiles.value}
        isReadOnly={isReadOnly}
      />
    </div>
  );
};

SubmissionContent.propTypes = {
  textResponses: PropTypes.shape({
    value: PropTypes.arrayOf(PropTypes.string).isRequired,
    onChange: PropTypes.func.isRequired,
    draftSaved: PropTypes.bool.isRequired,
  }).isRequired,
  uploadedFiles: PropTypes.shape({
    value: PropTypes.arrayOf(PropTypes.shape({
      fileDescription: PropTypes.string,
      fileName: PropTypes.string,
      fileSize: PropTypes.number,
    })),
    onDeletedFile: PropTypes.func.isRequired,
    onFileUploaded: PropTypes.func.isRequired,
  }).isRequired,
};

export default SubmissionContent;
