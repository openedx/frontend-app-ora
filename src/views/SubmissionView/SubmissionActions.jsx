import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from '@edx/frontend-platform/i18n';
import { ActionRow, StatefulButton } from '@edx/paragon';
import { MutationStatus } from 'data/services/lms/constants';
import messages from './messages';

const SubmissionActions = ({
  submitResponseHandler,
  submitResponseStatus,
  saveResponseHandler,
  saveResponseStatus,
}) => {
  const { formatMessage } = useIntl();

  return (
    <ActionRow>
      <StatefulButton
        variant="secondary"
        onClick={saveResponseHandler}
        state={saveResponseStatus}
        disabledStates={[MutationStatus.loading]}
        labels={{
          default: formatMessage(messages.saveActionSave),
          [MutationStatus.loading]: formatMessage(
            messages.saveActionSaving,
          ),
        }}
      />
      <StatefulButton
        onClick={submitResponseHandler}
        state={submitResponseStatus}
        disabledStates={[MutationStatus.loading, MutationStatus.success]}
        labels={{
          [MutationStatus.idle]: formatMessage(messages.submissionActionSubmit),
          [MutationStatus.loading]: formatMessage(
            messages.submissionActionSubmitting,
          ),
          [MutationStatus.success]: formatMessage(
            messages.submissionActionSubmitted,
          ),
        }}
      />
    </ActionRow>
  );
};

SubmissionActions.propTypes = {
  submitResponseHandler: PropTypes.func.isRequired,
  submitResponseStatus: PropTypes.oneOf(Object.values(MutationStatus))
    .isRequired,
  saveResponseHandler: PropTypes.func.isRequired,
  saveResponseStatus: PropTypes.oneOf(Object.values(MutationStatus))
    .isRequired,
};

export default SubmissionActions;
