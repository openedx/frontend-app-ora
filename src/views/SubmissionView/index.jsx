import React from 'react';

import { Col, Icon, Row } from '@edx/paragon';
import { CheckCircle } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import Rubric from 'components/Rubric';
import ModalActions from 'components/ModalActions';
import FileUpload from 'components/FileUpload';
import Instructions from 'components/Instructions';
import StatusAlert from 'components/StatusAlert';

import SubmissionPrompts from './SubmissionPrompts';
import useSubmissionViewData from './hooks';

import './index.scss';

import messages from './messages';

const SubmissionView = () => {
  const {
    actionOptions,
    showRubric,
    response: {
      textResponses,
      uploadedFiles,
    },
    onUpdateTextResponse,
    isDraftSaved,
    onDeletedFile,
    onFileUploaded,
    isReadOnly,
  } = useSubmissionViewData();

  const { formatMessage } = useIntl();

  const draftIndicator = (!isReadOnly && isDraftSaved) && (
    <p className="d-flex text-primary">
      <Icon src={CheckCircle} />
      {formatMessage(messages.draftSaved)}
    </p>
  );

  return (
    <div className="assessment-content-layout mr-auto ml-auto">
      <div className="content-wrapper">
        <Row className="flex-nowrap m-0">
          <Col className="p-0">
            <div>
              <div className="d-flex justify-content-between">
                <h1>{formatMessage(messages.yourResponse)}</h1>
                {draftIndicator}
              </div>

              <StatusAlert hasSubmitted={actionOptions.hasSubmitted} />
              <Instructions />
              <SubmissionPrompts {...{ textResponses, onUpdateTextResponse, isReadOnly }} />
              <FileUpload
                onDeletedFile={onDeletedFile}
                onFileUploaded={onFileUploaded}
                uploadedFiles={uploadedFiles}
                isReadOnly={isReadOnly}
              />
            </div>
            <ModalActions options={actionOptions} />
          </Col>
          {showRubric && <Rubric />}
        </Row>
      </div>
    </div>
  );
};

export default SubmissionView;
