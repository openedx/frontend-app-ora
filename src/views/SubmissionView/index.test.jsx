import { shallow } from '@edx/react-unit-test-utils';

import useSubmissionViewData from './hooks';

import SubmissionView from './index';

jest.mock('components/Rubric', () => 'Rubric');
jest.mock('components/ModalActions', () => 'ModalActions');
jest.mock('components/FileUpload', () => 'FileUpload');
jest.mock('components/Instructions', () => 'Instructions');
jest.mock('components/StatusAlert', () => 'StatusAlert');
jest.mock('./SubmissionPrompts', () => 'SubmissionPrompts');
jest.mock('./hooks', () => jest.fn());

describe('<SubmissionView />', () => {
  const mockUseSubmissionViewData = {
    actionOptions: {
      hasSubmitted: false,
    },
    showRubric: false,
    response: {
      textResponses: ['response1', 'response2'],
      uploadedFiles: [],
    },
    onUpdateTextResponse: jest.fn(),
    isDraftSaved: false,
    onDeletedFile: jest.fn(),
    onFileUploaded: jest.fn(),
    isReadOnly: false,
  };
  useSubmissionViewData.mockReturnValue(mockUseSubmissionViewData);

  it('renders disable show rubric', () => {
    const wrapper = shallow(<SubmissionView />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('Rubric').length).toBe(0);
  });

  it('renders enable show rubric', () => {
    mockUseSubmissionViewData.showRubric = true;
    const wrapper = shallow(<SubmissionView />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('Rubric').length).toBe(1);
  });
});
