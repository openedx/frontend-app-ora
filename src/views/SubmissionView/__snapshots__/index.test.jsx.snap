// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubmissionView /> renders disable show rubric 1`] = `
<div
  className="assessment-content-layout mr-auto ml-auto"
>
  <div
    className="content-wrapper"
  >
    <Row
      className="flex-nowrap m-0"
    >
      <Col
        className="p-0"
      >
        <div>
          <div
            className="d-flex justify-content-between"
          >
            <h1>
              Your response
            </h1>
          </div>
          <StatusAlert
            hasSubmitted={false}
          />
          <Instructions />
          <SubmissionPrompts
            isReadOnly={false}
            onUpdateTextResponse={[MockFunction]}
            textResponses={
              Array [
                "response1",
                "response2",
              ]
            }
          />
          <FileUpload
            isReadOnly={false}
            onDeletedFile={[MockFunction]}
            onFileUploaded={[MockFunction]}
            uploadedFiles={Array []}
          />
        </div>
        <ModalActions
          options={
            Object {
              "hasSubmitted": false,
            }
          }
        />
      </Col>
    </Row>
  </div>
</div>
`;

exports[`<SubmissionView /> renders enable show rubric 1`] = `
<div
  className="assessment-content-layout mr-auto ml-auto"
>
  <div
    className="content-wrapper"
  >
    <Row
      className="flex-nowrap m-0"
    >
      <Col
        className="p-0"
      >
        <div>
          <div
            className="d-flex justify-content-between"
          >
            <h1>
              Your response
            </h1>
          </div>
          <StatusAlert
            hasSubmitted={false}
          />
          <Instructions />
          <SubmissionPrompts
            isReadOnly={false}
            onUpdateTextResponse={[MockFunction]}
            textResponses={
              Array [
                "response1",
                "response2",
              ]
            }
          />
          <FileUpload
            isReadOnly={false}
            onDeletedFile={[MockFunction]}
            onFileUploaded={[MockFunction]}
            uploadedFiles={Array []}
          />
        </div>
        <ModalActions
          options={
            Object {
              "hasSubmitted": false,
            }
          }
        />
      </Col>
      <Rubric />
    </Row>
  </div>
</div>
`;
